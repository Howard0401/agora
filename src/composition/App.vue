<template>
  <div id="agora"> 
    <b-navbar toggleable="md" type="dark" variant="info">
      <b-navbar-toggle target="nav_collapse"></b-navbar-toggle>
      <b-navbar-brand href="/">Dashboard</b-navbar-brand>
    </b-navbar>
    <router-view/>
  </div>
</template>

<script lang='ts'>
import Vue from "vue"
import { Component } from "vue-property-decorator"

@Component
export default class App extends Vue { }
</script>
