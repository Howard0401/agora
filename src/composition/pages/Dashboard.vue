<template>
    <div class="dashboard container">
        <section>
            <h1>Dashboard</h1>
        </section>
        <section>
            <person-overview/>
        </section>
    </div>
</template>

<script lang='ts'>
import Vue from "vue"
import { Component } from "vue-property-decorator"

import PersonOverview from "@/persons/PersonOverview.vue"

@Component({
    components: {
        PersonOverview
    }
})
export default class AnalyticsPage extends Vue {
}
</script>
