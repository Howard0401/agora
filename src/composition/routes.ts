import { RouteConfig, Route } from "vue-router"

import Dashboard from "./pages/Dashboard.vue"

import PersonEdit from "@/persons/PersonEdit.vue"

export const routes: RouteConfig[] = [
    { path: "/", component: Dashboard },
    {
        path: "/persons/:id/edit",
        component: PersonEdit,
        props: (route: Route) => ({
            id: Number.parseInt(route.params.id)
        })
    },
]
