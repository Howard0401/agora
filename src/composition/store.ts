import { StoreOptions } from "vuex"

import * as persons from "@/persons/store"

export interface RootState {
    persons: persons.State
}

export const storeOptions: StoreOptions<RootState> = {
    modules: {
        persons: persons.module,
    },
    strict: process.env.NODE_ENV !== "production"
}
