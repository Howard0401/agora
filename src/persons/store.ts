import { Module, ActionContext } from "vuex"

import { RootState } from "@/composition/store"

import { Person } from "@/persons/model/Person"

export interface State {
    data: Person[]
}

export function initialState(): State {
    return {
        data: [
            new Person(1, "Peter Jackson"),
            new Person(2, "Elijah Wood"),
            new Person(3, "Orlando Bloom"),
            new Person(4, "Christopher Lee"),
        ],
    }
}

export const module: Module<State, RootState> = {
    state: initialState(),
    getters: {
        personById(state: State): (id: number) => Person | undefined {
            return id => state.data.find(person => person.id === id)
        },
        personIndexById(state: State): (id: number) => number {
            return id => state.data.findIndex(person => person.id === id)
        }
    },
    actions: {
        updatePerson(context: ActionContext<State, RootState>, person: Person): boolean {
            let success = false
            if (person.name.length > 0) {
                const index = context.getters.personIndexById(person.id)
                if (index >= 0 && person.isValid) {
                    context.commit("personUpdated", { index, person })
                    success = true
                }
            }
            return success
        }
    },
    mutations: {
        personUpdated(state: State, payload: { index: number, person: Person }): void {
            state.data[payload.index] = payload.person
        }
    },
    namespaced: true
}
