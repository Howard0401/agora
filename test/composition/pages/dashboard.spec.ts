import _ from "lodash"
import Vue from "vue"
import { Store, StoreOptions } from "vuex"

import { mount, Wrapper } from "@vue/test-utils"
import { MockStoreBuilder } from "../../test-utils/mock-store-builder"

import VueRouter from "vue-router"
import { routes } from "@/composition/routes"
const router = new VueRouter({ routes })

import { RootState, storeOptions } from "@/composition/store"
import Dashboard from "@/composition/pages/Dashboard.vue"

describe("Dashboard", () => {

    describe("no data loaded", () => {

        let page: Wrapper<Vue>
        let options: StoreOptions<RootState>
        let store: Store<RootState>
        beforeEach(() => {
            options = _.cloneDeep(storeOptions)
            store = new MockStoreBuilder<RootState>(options)
                .build()
            page = mount(Dashboard, { store, router })
        })

        it("renders an empty page", () => {
            expect(page.element.outerHTML).toMatchSnapshot()
        })

        it("renders the title", () => {
            expect(page.find("section").text()).toEqual("Dashboard")
        })
    })
})
