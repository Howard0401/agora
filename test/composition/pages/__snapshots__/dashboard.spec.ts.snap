// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard no data loaded renders an empty page 1`] = `
<div class="dashboard container">
    <section>
        <h1>Dashboard</h1>
    </section>
    <section>
        <div id="person-list">
            <div id="person-list">
                <div id="person" class="card">
                    <div class="card-body">
                        <h4 class="card-title">Peter Jackson</h4><a href="#/persons/1/edit" class="">edit</a></div>
                </div>
                <div id="person" class="card">
                    <div class="card-body">
                        <h4 class="card-title">Elijah Wood</h4><a href="#/persons/2/edit" class="">edit</a></div>
                </div>
                <div id="person" class="card">
                    <div class="card-body">
                        <h4 class="card-title">Orlando Bloom</h4><a href="#/persons/3/edit" class="">edit</a></div>
                </div>
                <div id="person" class="card">
                    <div class="card-body">
                        <h4 class="card-title">Christopher Lee</h4><a href="#/persons/4/edit" class="">edit</a></div>
                </div>
            </div>
        </div>
    </section>
</div>
`;
